File: generated_code/GeminiPro/Python/022_SplayTree.py
class SplayTree:
    """
    A self-contained Python class implementing a splay tree.
    This tree functions as a dictionary-like set for storing unique integers.
    The main operations (insert, delete, search) are implemented with the
    characteristic splaying operation to maintain amortized logarithmic time
    complexity.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty splay tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splaying operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            else:  # Zig-Zag case
                if x == parent.right and parent == grandparent.left:
                    self._left_rotate(parent)
                    self._right_rotate(grandparent)
                else: # x == parent.left and parent == grandparent.right
                    self._right_rotate(parent)
                    self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        If the key is found, the corresponding node is splayed to the root and
        True is returned.
        
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root and False is returned.
        
        Returns:
            bool: True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_node = None
        while current:
            last_node = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Key not found, splay the last accessed node
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and return
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # Insert new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.
        
        The search for the key will cause a splay operation. If the key is
        found, it is removed, and its subtrees are merged.
        """
        # First, search for the key. This will splay it to the root if found.
        if not self.search(key):
            # Key not in the tree, nothing to delete.
            # search() already splayed the closest node.
            return

        # At this point, the node to delete is the root
        node_to_delete = self.root
        
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            # No left child, the right subtree becomes the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect left subtree from the node to be deleted
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this maximum element to the root of the left subtree
            # Temporarily set the root to perform splay within the subtree
            self.root = left_subtree
            self._splay(max_node)

            # The new root (max_node) now has the right_subtree as its right child
            max_node.right = right_subtree
            if right_subtree:
                right_subtree.parent = max_node
            
            self.root = max_node

File: generated_code/GeminiPro/Java/022_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree that stores integers.
 * Splay trees are self-balancing binary search trees with the property that
 * recently accessed elements are quick to access again. This implementation
 * supports insertion, deletion, and search operations.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     * It includes a parent pointer to simplify rotation and splay operations.
     */
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
            this.parent = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splay Operations ---

    /**
     * Performs a left rotation on the given node x.
     *
     *      x           y
     *     / \         / \
     *    A   y  -->  x   C
     *       / \     / \
     *      B   C   A   B
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node y.
     *
     *        y           x
     *       / \         / \
     *      x   C  -->  A   y
     *     / \             / \
     *    A   B           B   C
     *
     * @param y The node to rotate.
     */
    private void rightRotate(Node y) {
        Node x = y.left;
        if (x != null) {
            y.left = x.right;
            if (x.right != null) {
                x.right.parent = y;
            }
            x.parent = y.parent;
            if (y.parent == null) {
                this.root = x;
            } else if (y == y.parent.left) {
                y.parent.left = x;
            } else {
                y.parent.right = x;
            }
            x.right = y;
            y.parent = x;
        }
    }

    /**
     * Performs the splay operation on a node x, moving it to the root.
     * This is done by a series of rotations (Zig, Zig-Zig, Zig-Zag).
     *
     * @param x The node to splay to the root.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig case
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else {
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig case
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag case
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else {
                    if (parent == grandParent.right) { // Zig-Zig case
                        leftRotate(grandParent);
                        leftRotate(parent);
                    } else { // Zig-Zag case
                        leftRotate(parent);
                        rightRotate(grandParent);
                    }
                }
            }
        }
    }

    // --- Public API ---

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the existing node is splayed to the root.
     * Otherwise, a new node is created and splayed to the root.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay this node and return.
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;

        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        // Splay the newly inserted node to the root.
        splay(newNode);
    }

    /**
     * Searches for a key in the tree and performs the splaying operation.
     * This method will splay the node containing the key to the root if found.
     * If the key is not found, it splays the last accessed node on the search path.
     *
     * @param key The key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        
        Node current = root;
        Node lastNode = root; // Keep track of the last non-null node on the path
        
        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key found, splay the node and return
                splay(current);
                return current.key;
            }
        }
        
        // Key not found, splay the last visited node
        splay(lastNode);
        
        // After splaying, check if the new root has the key
        if (root.key == key) {
            return root.key;
        } else {
            return null;
        }
    }

    /**
     * Deletes a key from the splay tree.
     * This operation first finds the node, splays it to the root, and then removes it
     * by joining its two subtrees.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // Search for the key. This will bring the node (or a nearby one) to the root.
        search(key);

        // If the key is not at the root after searching, it was not in the tree.
        if (root == null || root.key != key) {
            return;
        }

        // The node to delete is now at the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            // Disconnect the left subtree to work on it.
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree.
            Node maxInLeft = leftSubtree;
            while (maxInLeft.right != null) {
                maxInLeft = maxInLeft.right;
            }

            // Splay this max node to the root of the left subtree.
            // Temporarily set the main root to the left subtree's root to use the splay method.
            this.root = leftSubtree;
            splay(maxInLeft);

            // After splaying, `this.root` is the new root (maxInLeft).
            // Now, join the original right subtree.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}